app/uk/gov/hmrc/tai/model/hip/reads/CodingComponentHipReads.scala

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright 2024 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.tai.model.hip.reads
+
+import play.api.libs.json.*
+import uk.gov.hmrc.tai.model.domain.*
+import uk.gov.hmrc.tai.model.domain.calculation.CodingComponent
+import uk.gov.hmrc.tai.model.hip.reads.NpsIabdSummaryHipReads.iabdsFromTotalLiabilityReads
+import uk.gov.hmrc.tai.util.JsonHelper.parseTypeOrException
+
+object CodingComponentHipReads {
+
+  private type CodingComponentFactory = (Int, BigDecimal, String, Option[BigDecimal]) => Option[CodingComponent]
```

Comment:
This is better than how it was. However, I'd personally prefer it if we got rid of these factory vals and instead use the json transformation syntax, e.g. (JsPath \ "deductionsDetails").read[Seq[CodingComponent]](readsDeductionDetail) instead of extractComponents(... deductionFactory). This would, to my mind, be cleaner and more readable. Just my opinion though, what do you think?

Reply:
Agree with Grant. I think it would be more readable.

Reply:
Will change it.

Reply:
Looks much better now